java/com/google/flatbuffers/Table.java

```diff
@@ -71,20 +83,34 @@ protected int __indirect(int offset) {
    * @return Returns a `String` from the data stored inside the FlatBuffer at `offset`.
    */
   protected String __string(int offset) {
+    CharsetDecoder decoder = UTF8_DECODER.get();
+    decoder.reset();
+
     offset += bb.getInt(offset);
-    if (bb.hasArray()) {
-      return new String(bb.array(), bb.arrayOffset() + offset + SIZEOF_INT, bb.getInt(offset),
-                        FlatBufferBuilder.utf8charset);
-    } else {
-      // We can't access .array(), since the ByteBuffer is read-only,
-      // off-heap or a memory map
-      ByteBuffer bb = this.bb.duplicate().order(ByteOrder.LITTLE_ENDIAN);
-      // We're forced to make an extra copy:
-      byte[] copy = new byte[bb.getInt(offset)];
-      bb.position(offset + SIZEOF_INT);
-      bb.get(copy);
-      return new String(copy, 0, copy.length, FlatBufferBuilder.utf8charset);
+    ByteBuffer src = bb.duplicate().order(ByteOrder.LITTLE_ENDIAN);
+    int length = src.getInt(offset);
+    src.position(offset + SIZEOF_INT);
+    src.limit(offset + SIZEOF_INT + length);
+
+    int required = (int)((float)length * decoder.maxCharsPerByte());
+    CharBuffer dst = CHAR_BUFFER.get();
+    if (dst == null || dst.capacity() < required) {
+      dst = CharBuffer.allocate(Math.max(required, 128));
```

Comment:
if this happens on the main thread of a program, that data will be allocated till the end. But yes, you're right, it is shared for all tables.

I mean, don't have a minimum of 128 bytes even. Let the minimum be the size of the first string. Saves memory in the case of small strings, and shouldn't hurt case that read lots of strings.